cmake/Util.cmake

```diff
@@ -0,0 +1,145 @@
+# target, then the target will only run its commands for file A and ignore
+# file B.
+#
+# Another example: let's say we have file A and B, but now we create only a
+# single touch file T for both of them. This would mean that if I change
+# either file A or B, then the target will run its commands on both A and B.
+# Meaning that even if I only change file A, the target will still run
+# commands on both A and B.
+#
+# The more touch files we create for a target, the fewer commands we'll need
+# to rerun, and by extension, the more time we'll save. Unfortunately, the
+# more touch files we create the more intermediary targets will be created,
+# one for each touch file. This makes listing all targets with
+# `cmake --build build --target help` less useful since each touch file will
+# be listed. The tradeoff that needs to be done here is between performance
+# and "discoverability". As a general guideline: the more popular a target is
+# and the more time it takes to run it, the more granular you want your touch
+# files to be. Conversely, if a target rarely needs to be run or if it's fast,
+# then you should create fewer targets.
+#
+# Possible values for TOUCH_STRATEGY:
+# "SINGLE":   create a single touch file for all files.
+# "PER_FILE": create a touch file for each file. Defaults to this if
+#             TOUCH_STRATEGY isn't specified.
+# "PER_DIR":  create a touch file for each directory.
+#
+# List arguments:
+# FLAGS     - List of flags to use after COMMAND
+# FILES     - List of files to use COMMAND on. It's possible to combine this
+#             with GLOB_PAT and GLOB_DIRS; the files found by globbing will
+#             simple be added to FILES
+# GLOB_DIRS - The directories to recursively search for files with extension
+#             GLOB_PAT
+#
+function(add_glob_targets)
+  cmake_parse_arguments(ARG
+    "REQUIRED"
+    "TARGET;COMMAND;GLOB_PAT;TOUCH_STRATEGY"
+    "FLAGS;FILES;GLOB_DIRS"
+    ${ARGN}
+  )
+
+  if(NOT ARG_COMMAND)
+    add_custom_target(${ARG_TARGET})
+    if(ARG_REQUIRED)
+      add_custom_command(TARGET ${ARG_TARGET}
+        COMMAND ${CMAKE_COMMAND} -E echo "${ARG_TARGET}: ${ARG_COMMAND} not found"
+        COMMAND false)
+    else()
+      add_custom_command(TARGET ${ARG_TARGET}
+        COMMAND ${CMAKE_COMMAND} -E echo "${ARG_TARGET} SKIP: ${ARG_COMMAND} not found")
+    endif()
+    return()
+  endif()
+
+  foreach(gd ${ARG_GLOB_DIRS})
+    file(GLOB_RECURSE globfiles ${PROJECT_SOURCE_DIR}/${gd}/${ARG_GLOB_PAT})
+    list(APPEND ARG_FILES ${globfiles})
+  endforeach()
+
+  if(NOT ARG_TOUCH_STRATEGY)
+    set(ARG_TOUCH_STRATEGY PER_FILE)
+  endif()
+  set(POSSIBLE_TOUCH_STRATEGIES SINGLE PER_FILE PER_DIR)
+  if(NOT ARG_TOUCH_STRATEGY IN_LIST POSSIBLE_TOUCH_STRATEGIES)
+    message(FATAL_ERROR "Unrecognized value for TOUCH_STRATEGY: ${ARG_TOUCH_STRATEGY}")
+  endif()
+
+  if(ARG_TOUCH_STRATEGY STREQUAL SINGLE)
+    set(touch_file ${TOUCHES_DIR}/${ARG_TARGET})
```

Comment:
in this touch files, should we use the `ran-` prefix as we've done elsewhere? That's at least helpful as a hint when browsing `build/` I guess.

Reply:
Eh, feels a bit redundant since it's already under the `touches` directory, but I don't mind it.